File: go/pricing/applieddiscount.go
// ABOUTME: AppliedDiscount domain type representing a discount applied to a cart or line item.
// ABOUTME: Preserves explainability with promotion ID, amount, target, details, and optional allocations.
package pricing

import "promotionengine/domain"

// AppliedDiscount represents a discount that has been applied.
// It preserves explainability by tracking the promotion, amount, target, and details.
type AppliedDiscount struct {
	PromotionID domain.PromotionID
	Amount      domain.Money
	Target      string // "line" or "cart"
	Details     string
	Allocations map[domain.Sku]domain.Money // Optional: per-SKU discount allocations
}

// NewAppliedDiscount creates a new AppliedDiscount without allocations.
func NewAppliedDiscount(promotionID domain.PromotionID, amount domain.Money, target, details string) AppliedDiscount {
	return AppliedDiscount{
		PromotionID: promotionID,
		Amount:      amount,
		Target:      target,
		Details:     details,
		Allocations: nil,
	}
}

// NewAppliedDiscountWithAllocations creates a new AppliedDiscount with per-SKU allocations.
func NewAppliedDiscountWithAllocations(promotionID domain.PromotionID, amount domain.Money, target, details string, allocations map[domain.Sku]domain.Money) AppliedDiscount {
	return AppliedDiscount{
		PromotionID: promotionID,
		Amount:      amount,
		Target:      target,
		Details:     details,
		Allocations: allocations,
	}
}

File: typescript/src/pricing/AppliedDiscount.ts
// ABOUTME: Record of a discount that was applied to a cart
// ABOUTME: Includes promotion ID, amount, target scope, details, and optional allocations

import { PromotionId } from '../domain/PromotionId';
import { Money } from '../domain/Money';
import { Sku } from '../domain/Sku';

export class AppliedDiscount {
  constructor(
    public readonly promotionId: PromotionId,
    public readonly amount: Money,
    public readonly target: 'line' | 'cart',
    public readonly details: string,
    public readonly allocations?: ReadonlyMap<Sku, Money>
  ) {}
}
